File: lightroom_tagger/core/description_service.py
"""Describe matched catalog and Instagram images on demand."""
import os
import sqlite3
import threading
from collections.abc import Callable
from typing import TypedDict

from lightroom_tagger.core.analyzer import VIDEO_EXTENSIONS, build_description_op_spec
from lightroom_tagger.core.config import get_description_model
from lightroom_tagger.core.database import (
    get_image,
    get_image_description,
    get_instagram_dump_media,
    list_perspectives,
    resolve_filepath,
    store_image_description,
)
from lightroom_tagger.core.prompt_builder import build_description_user_prompt
from lightroom_tagger.core.vision_cache import get_or_create_cached_image
from lightroom_tagger.core.vision_op import VisionOpOutcome, run_vision_op_persist

LogCallback = Callable[[str, str], None] | None


class DescribeTelemetry(TypedDict):
    """Thread-safe telemetry bag for batch describe passes.

    Construct with::

        telemetry: DescribeTelemetry = {
            'silent_compression_skips': 0,
            '_lock': threading.Lock(),
        }

    Pass ``None`` when telemetry is not needed; any partial dict raises KeyError
    at the ``with telemetry['_lock']`` site.
    """

    silent_compression_skips: int
    _lock: threading.Lock


# File extensions the vision pipeline cannot describe. These get short-circuited
# before we ever call into compression / the provider dispatcher — otherwise a
# .mov file falls all the way through to the LLM (compress_image silently
# returns the raw bytes on failure) and stalls the worker on multi-minute retry
# backoffs, which in turn wedges the batch-describe ``as_completed`` loop.
_NON_DESCRIBABLE_EXTENSIONS = VIDEO_EXTENSIONS


def _is_non_describable_path(filepath: str | None) -> bool:
    if not filepath:
        return False
    ext = os.path.splitext(filepath)[1].lower()
    return ext in _NON_DESCRIBABLE_EXTENSIONS


def _description_structured_is_valid(structured: dict) -> bool:
    """True if vision-op description output is worth persisting (non-empty summary)."""
    summary = structured.get('summary')
    return isinstance(summary, str) and bool(summary.strip())


def _resolve_description_user_prompt(
    db: sqlite3.Connection, perspective_slugs: list[str] | None
) -> str | None:
    """Return assembled user prompt text, or ``None`` to use the legacy monolithic prompt."""
    active_rows = list_perspectives(db, active_only=True)
    if perspective_slugs:
        by_slug = {r["slug"]: r for r in active_rows}
        rows = [by_slug[s] for s in perspective_slugs if s in by_slug]
    else:
        rows = active_rows
    if not rows:
        return None
    return build_description_user_prompt(rows)


def _persist_description_structured(
    db: sqlite3.Connection,
    image_key: str,
    image_type: str,
    structured: dict,
    provider: str,
    model: str,
) -> None:
    model_used = structured.pop("_model", None) or model or get_description_model()
    provider_used = structured.pop("_provider", None) or provider
    model_label = f"{provider_used}:{model_used}" if provider_used else model_used
    _store_structured(db, image_key, image_type, structured, model_label)


def _run_description_persist(
    db: sqlite3.Connection,
    image_key: str,
    image_type: str,
    image_path: str,
    *,
    provider_id: str | None,
    model: str | None,
    log_callback: LogCallback,
    user_prompt: str | None,
    silent_compression: bool,
) -> VisionOpOutcome:
    spec = build_description_op_spec(
        image_path,
        provider_id=provider_id,
        model=model,
        log_callback=log_callback,
        user_prompt=user_prompt,
        silent_compression=silent_compression,
    )

    def persist(structured: dict, provider: str, model_used: str) -> None:
        structured = dict(structured)
        structured["_provider"] = provider
        structured["_model"] = model_used
        _persist_description_structured(db, image_key, image_type, structured, provider, model_used)

    return run_vision_op_persist(
        spec,
        accept_result=_description_structured_is_valid,
        persist=persist,
    )


def describe_matched_image(db: sqlite3.Connection, catalog_key: str, force: bool = False,
                           provider_id: str | None = None,
                           model: str | None = None,
                           log_callback: LogCallback = None,
                           perspective_slugs: list[str] | None = None,
                           *,
                           telemetry: DescribeTelemetry | None = None) -> VisionOpOutcome:
    """Generate and store a description for a catalog image if needed.

    Returns a :class:`VisionOpOutcome`. ``wrote`` is True when a non-empty
    description was stored. Skipped or failed outcomes leave the DB unchanged.
    """
    if not force and get_image_description(db, catalog_key):
        return VisionOpOutcome(status='skipped', reason='description exists')

    image = get_image(db, catalog_key)
    if not image or not image.get('filepath'):
        return VisionOpOutcome(status='skipped', reason='image not found')

    filepath = resolve_filepath(image['filepath'])
    if _is_non_describable_path(filepath):
        return VisionOpOutcome(status='skipped', reason='non-describable file type')
    if not os.path.exists(filepath):
        return VisionOpOutcome(status='skipped', reason='file missing')

    cached_path = get_or_create_cached_image(db, catalog_key, filepath)
    image_for_describe = cached_path if cached_path and os.path.exists(cached_path) else filepath
    user_prompt = _resolve_description_user_prompt(db, perspective_slugs)
    use_silent_compression = (
        bool(cached_path)
        and cached_path == image_for_describe
        and os.path.exists(cached_path)
    )

    outcome = _run_description_persist(
        db,
        catalog_key,
        'catalog',
        image_for_describe,
        provider_id=provider_id,
        model=model,
        log_callback=log_callback,
        user_prompt=user_prompt,
        silent_compression=use_silent_compression,
    )
    if use_silent_compression and telemetry is not None:
        with telemetry['_lock']:
            telemetry['silent_compression_skips'] += 1
    return outcome


def describe_instagram_image(db: sqlite3.Connection, media_key: str, force: bool = False,
                             provider_id: str | None = None,
                             model: str | None = None,
                             log_callback: LogCallback = None,
                             perspective_slugs: list[str] | None = None) -> VisionOpOutcome:
    """Generate and store a description for an Instagram image if needed.

    Uses the local file from instagram_dump_media. ``wrote`` is True when a
    non-empty description was stored.
    """
    if not force and get_image_description(db, media_key):
        return VisionOpOutcome(status='skipped', reason='description exists')

    dump_media = get_instagram_dump_media(db, media_key)
    if not dump_media or not dump_media.get('file_path'):
        return VisionOpOutcome(status='skipped', reason='image not found')

    filepath = dump_media['file_path']
    if _is_non_describable_path(filepath):
        return VisionOpOutcome(status='skipped', reason='non-describable file type')
    if not os.path.exists(filepath):
        return VisionOpOutcome(status='skipped', reason='file missing')

    user_prompt = _resolve_description_user_prompt(db, perspective_slugs)
    return _run_description_persist(
        db,
        media_key,
        'instagram',
        filepath,
        provider_id=provider_id,
        model=model,
        log_callback=log_callback,
        user_prompt=user_prompt,
        silent_compression=False,
    )


def _store_structured(
    db: sqlite3.Connection,
    image_key: str,
    image_type: str,
    structured: dict,
    model_used: str | None = None,
) -> None:
    technical = structured.get('technical')
    if not isinstance(technical, dict):
        technical = {}

    raw_dc = structured.get('dominant_colors')
    if isinstance(raw_dc, list) and len(raw_dc) > 0:
        dc: list | None = raw_dc
    else:
        tc = technical.get('dominant_colors')
        dc = tc if isinstance(tc, list) else None

    raw_mt = structured.get('mood_tags')
    if isinstance(raw_mt, list):
        mt: list | None = raw_mt
    else:
        mt = None
    if mt is None:
        mood = technical.get('mood')
        if isinstance(mood, str) and mood.strip():
            mt = [mood.strip()]

    hr_raw = structured.get('has_repetition')
    if hr_raw is None:
        hr = None
    elif isinstance(hr_raw, bool):
        hr = hr_raw
    elif isinstance(hr_raw, int) and hr_raw in (0, 1):
        hr = hr_raw
    else:
        hr = hr_raw

    store_image_description(db, {
        'image_key': image_key,
        'image_type': image_type,
        'summary': structured.get('summary', ''),
        'composition': structured.get('composition', {}),
        'perspectives': structured.get('perspectives', {}),
        'technical': structured.get('technical', {}),
        'subjects': structured.get('subjects', []),
        'best_perspective': structured.get('best_perspective', ''),
        'model_used': model_used or get_description_model(),
        'dominant_colors': dc,
        'mood_tags': mt,
        'has_repetition': hr,
    })

File: apps/visualizer/backend-ts/src/vision/description-service.ts
/**
 * Describe catalog images on demand.
 */
import { extname } from 'node:path';
import { getDescriptionModel } from '../config.js';
import { buildDescriptionOpSpec, type DescriptionStructured } from '../analyzer/description.js';
import { VIDEO_EXTENSIONS } from '../imaging/raw-decode.js';
import type { ProviderRegistry } from '../providers/registry.js';
import type { CancelCheck, LogCallback } from '../providers/retry.js';
import type { ConsecutiveAbortTracker, ErrorPolicy } from '../providers/error-policy.js';
import { buildDescriptionUserPrompt } from '../analyzer/prompt-builder.js';
import { getImage } from '../db/library/catalog.js';
import { getImageDescription, storeImageDescription } from '../db/library/descriptions.js';
import { libraryWrite } from '../db/library/write.js';
import type { Db } from '../db/connection.js';
import { resolveFilepath } from '../utils/path-resolve.js';
import { resolveVisionImage } from './vision-cache.js';
import { runVisionOpPersist, VisionOpOutcome } from './vision-op.js';

/**
 * Counters for a batch describe pass.
 *
 * Python needs a `threading.Lock` in this bag because its describe workers are
 * real threads sharing one interpreter. Here the batch handler drives concurrency
 * with promises on a single isolate, so `silentCompressionSkips += 1` cannot
 * interleave and no lock is needed — the field is the whole struct.
 */
export interface DescribeTelemetry {
  silentCompressionSkips: number;
}

/**
 * Extensions the vision pipeline cannot describe, short-circuited before
 * compression or dispatch.
 *
 * Without this a `.mov` falls all the way through to the LLM — `compressImage`
 * silently returns the original path when it cannot decode — and then stalls the
 * worker on multi-minute retry backoffs, which wedges the whole batch.
 */
const NON_DESCRIBABLE_EXTENSIONS = VIDEO_EXTENSIONS;

function isNonDescribablePath(filepath: string | null | undefined): boolean {
  if (!filepath) return false;
  return NON_DESCRIBABLE_EXTENSIONS.has(extname(filepath).toLowerCase());
}

/**
 * True when the op's output is worth storing.
 *
 * A blank summary is indistinguishable from a real description in the UI, and
 * storing one would stop the image ever being retried — so an empty answer must
 * fail rather than persist.
 */
export function descriptionStructuredIsValid(structured: DescriptionStructured): boolean {
  const summary = structured.summary;
  return typeof summary === 'string' && summary.trim().length > 0;
}

/**
 * Normalize the model's loose output into the columns `image_descriptions` holds.
 *
 * Every fallback here exists because models answer the prompt's root-level
 * `dominant_colors` / `mood_tags` / `has_repetition` fields inconsistently, and
 * the same values are also asked for inside `technical`. The order is Python's:
 *
 *   - `dominant_colors`: root list if non-empty, else `technical.dominant_colors`.
 *     An empty root list falls through, because "the model returned []" and "the
 *     model omitted it" mean the same thing for indexing.
 *   - `mood_tags`: root list if it *is* a list — including `[]`, which is taken as
 *     a deliberate "no tags" and does not fall through — else `[technical.mood]`.
 *   - `has_repetition`: passed through untouched, including non-boolean junk, and
 *     coerced by `storeImageDescription`.
 */
export function storeStructured(
  db: Db,
  imageKey: string,
  imageType: string,
  structured: DescriptionStructured,
  modelUsed: string | null = null,
): void {
  const rawTechnical = structured.technical;
  const technical: Record<string, unknown> =
    rawTechnical !== null && typeof rawTechnical === 'object' && !Array.isArray(rawTechnical)
      ? (rawTechnical as Record<string, unknown>)
      : {};

  const rawDc = structured.dominant_colors;
  let dc: unknown[] | null;
  if (Array.isArray(rawDc) && rawDc.length > 0) {
    dc = rawDc;
  } else {
    const tc = technical.dominant_colors;
    dc = Array.isArray(tc) ? tc : null;
  }

  const rawMt = structured.mood_tags;
  let mt: unknown[] | null = Array.isArray(rawMt) ? rawMt : null;
  if (mt === null) {
    const mood = technical.mood;
    if (typeof mood === 'string' && mood.trim()) mt = [mood.trim()];
  }

  const hr = structured.has_repetition ?? null;

  libraryWrite(db, () =>
    storeImageDescription(db, {
      image_key: imageKey,
      image_type: imageType,
      summary: typeof structured.summary === 'string' ? structured.summary : '',
      composition: structured.composition ?? {},
      technical: structured.technical ?? {},
      subjects: structured.subjects ?? [],
      model_used: modelUsed || getDescriptionModel(),
      dominant_colors: dc,
      mood_tags: mt,
      has_repetition: hr,
    }),
  );
}

export interface DescribeOptions {
  force?: boolean;
  providerId?: string | null;
  model?: string | null;
  logCallback?: LogCallback;
  registry?: ProviderRegistry | null;
  errorPolicy?: ErrorPolicy | null;
  abortTracker?: ConsecutiveAbortTracker | null;
  cancelCheck?: CancelCheck;
  telemetry?: DescribeTelemetry | null;
}

/**
 * Generate and store a description for one catalog image.
 *
 * `wrote` is true only when a non-empty description was stored; skipped and
 * failed outcomes leave the database untouched.
 */
export async function describeMatchedImage(
  db: Db,
  catalogKey: string,
  opts: DescribeOptions = {},
): Promise<VisionOpOutcome> {
  if (!opts.force && getImageDescription(db, catalogKey)) {
    return new VisionOpOutcome('skipped', 'description exists');
  }

  const image = getImage(db, catalogKey);
  const rawPath = image?.['filepath'];
  if (!image || typeof rawPath !== 'string' || !rawPath) {
    return new VisionOpOutcome('skipped', 'image not found');
  }

  const filepath = resolveFilepath(rawPath);
  if (isNonDescribablePath(filepath)) {
    return new VisionOpOutcome('skipped', 'non-describable file type');
  }

  // Prefer the local vision cache: the original may be unreachable — an unmounted
  // NAS is the normal case — and describe can still run off the cached JPEG.
  const { path: imageForDescribe, silentCompression } = await resolveVisionImage(
    db,
    catalogKey,
    filepath,
  );
  if (imageForDescribe === null) {
    return new VisionOpOutcome('skipped', 'file missing');
  }

  const spec = buildDescriptionOpSpec(imageForDescribe, {
    providerId: opts.providerId ?? null,
    model: opts.model ?? null,
    logCallback: opts.logCallback ?? null,
    userPrompt: buildDescriptionUserPrompt(),
    silentCompression,
    registry: opts.registry ?? null,
    cancelCheck: opts.cancelCheck ?? null,
  });
  spec.errorPolicy = opts.errorPolicy ?? null;
  spec.abortTracker = opts.abortTracker ?? null;

  const outcome = await runVisionOpPersist(spec, {
    acceptResult: descriptionStructuredIsValid,
    persist: (structured, provider, modelUsed) => {
      // `provider:model` is what the UI shows and what `desc_model` is filtered
      // on, so the provider half is included whenever it is known.
      const resolvedModel = modelUsed || opts.model || getDescriptionModel();
      const label = provider ? `${provider}:${resolvedModel}` : resolvedModel;
      storeStructured(db, catalogKey, 'catalog', structured, label);
    },
  });

  if (silentCompression && opts.telemetry) {
    opts.telemetry.silentCompressionSkips += 1;
  }
  return outcome;
}
